tool/src/dart/class.rs

```diff
@@ -0,0 +1,914 @@
+                } else {
+                    vec![]
+                };
+
+                let mut set_slice_conversions = Vec::new();
+
+                let set_expressions = self
+                    .gen_dart_to_c_for_type(&field.ty, name.clone(), &mut set_slice_conversions)
+                    .into_iter()
+                    .map(
+                        |PartiallyNamedExpression { suffix, expression }| NamedExpression {
+                            name: format!("this._underlying.{name}{suffix}").into(),
+                            expression,
+                        },
+                    )
+                    .collect();
+
+                FieldInfo {
+                    name,
+                    annotation,
+                    ffi_cast_type_name,
+                    dart_type_name,
+                    get_expression,
+                    set_cleanups,
+                    set_slice_conversions,
+                    set_expressions,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let methods = ty
+            .methods
+            .iter()
+            .flat_map(|method| self.gen_method_info(id, method, &type_name))
+            .collect::<Vec<_>>();
+
+        ImplTemplate {
+            type_name: &type_name,
+            fields: &fields,
+            methods: methods.as_slice(),
+            docs: self.cx.formatter.fmt_docs(&ty.docs),
+        }
+        .render()
+        .unwrap()
+    }
+
+    fn gen_method_info(
+        &mut self,
+        id: TypeId,
+        method: &'tcx hir::Method,
+        type_name: &str,
+    ) -> Option<MethodInfo<'dartcx>> {
+        if method.attrs.disable {
+            return None;
+        }
+
+        let _guard = self.cx.errors.set_context_method(
+            self.cx.formatter.fmt_type_name_diagnostics(id),
+            method.name.as_str().into(),
+        );
+
+        self.imports.insert(Import {
+            path: "dart:ffi".into(),
+            suffix: " as ffi".into(),
+        });
+
+        let c_method_name = self.cx.formatter.fmt_c_method_name(id, method);
+
+        let mut param_decls_dart = Vec::new();
+        let mut param_types_ffi = Vec::new();
+        let mut param_types_ffi_cast = Vec::new();
+        let mut dart_to_ffi_params = Vec::new();
+
+        if let Some(param_self) = method.param_self.as_ref() {
+            param_types_ffi.push(self.gen_self_type_ffi(&param_self.ty, false));
+            param_types_ffi_cast.push(self.gen_self_type_ffi(&param_self.ty, true));
+            dart_to_ffi_params.push(self.gen_dart_to_c_self(&param_self.ty));
+        }
+
+        let mut slice_conversions = Vec::new();
+
+        for param in method.params.iter() {
+            param_decls_dart.push(self.gen_ty_decl(&param.ty, param.name.as_str()));
+            let param_type_ffi = self.gen_type_name_ffi(&param.ty, false);
+            if let Some(element_type_ffi) = param_type_ffi.strip_prefix("Slice:") {
```

Comment:
nit: not sure I like this.

Reply:
Slice and string are the only types that get turned into two arguments. This allows abstracting most of that away, otherwise I'll always have to use variable length thingies for all types.

Reply:
can we have `gen_type_name_ffi` return a tuple or something, then?

Reply:
No, because when the slice is a field it should actually use the `Slice` type, and I don't want a separate `gen_type_name_ffi_for_field`